### ISSUE TITLE
Flink cannot restart from checkpoint due to shifted index order in DataFile

### ISSUE THREAD

**Issue by jackye1995** [apache/iceberg#3700]
track the issue found in https://github.com/apache/iceberg/pull/3015/files#r753631713 as 0.13 blocker

@stevenzwu @rdblue @RussellSpitzer @szehon-ho

**Comment by jackye1995** [comment]
I am not totally familiar with the PR, after reading it just now, it seems like the quickest fix is to move the `spec_id` column back as the last column, given this is not released yet in past versions. Would that be enough? @rdblue

**Comment by openinx** [comment]
I think I missed the previous comment from @stevenzwu  before, sorry for that. recently, I have been busy with some internal affairs from the company and don't have the enough bandwidth to check all the maillist and github notifications. 

I don't see the root cause when I initially check the whole flink manifest read/write path, I will need to reproduce this bug in my localhost to locate the root cause.  Will update this issue if there is any progress. 

Thanks.

**Comment by rdblue** [comment]
@jackye1995, a way to "fix" the problem is to move `spec_id` so that the original column ordinals match. The problem with doing that is that we leave a known bug in the codebase. And it's very unlikely that we'll get lucky and have someone catch it before release next time. There was no test that caught this.

Another reason not to is that we state that Iceberg metadata files follow Iceberg schema evolution rules. If we violate that here and start avoiding certain valid changes because of fear something might break, then we're back to Hadoop usability: you have to know what not to do and you're artificially limited. We don't want Iceberg developers to be afraid of using tools that are perfectly valid according to the spec.

**Comment by jackye1995** [comment]
Had some discussion with Steven and Ryan in Slack, the Flink checkpoint serialization is handled by `FlinkManifestUtil`, which calls the core library for serdes:

```java
  static ManifestFile writeDataFiles(OutputFile outputFile, PartitionSpec spec, List<DataFile> dataFiles)
      throws IOException {
    ManifestWriter<DataFile> writer = ManifestFiles.write(FORMAT_V2, spec, outputFile, DUMMY_SNAPSHOT_ID);

    try (ManifestWriter<DataFile> closeableWriter = writer) {
      closeableWriter.addAll(dataFiles);
    }

    return writer.toManifestFile();
  }

  static List<DataFile> readDataFiles(ManifestFile manifestFile, FileIO io) throws IOException {
    try (CloseableIterable<DataFile> dataFiles = ManifestFiles.read(manifestFile, io)) {
      return Lists.newArrayList(dataFiles);
    }
  }
```

The schema at write time is fixed through `MetadataV1/2` even when we add new fields in `DataFile`, so there should not be any compatibility issue.

I tried the following using 0.12 and 0.13:

In 0.12:

```
OutputFile outputFile = icebergTable.io().newOutputFile("file:/Users/yzhaoqin/0.12-out.avro");
FlinkManifestUtil.writeDataFiles(outputFile, icebergTable.spec(), Lists.newArrayList(icebergTable.currentSnapshot().addedFiles()));
```

In 0.13:

```
List<DataFile> dataFiles = FlinkManifestUtil.readDataFiles(
    new GenericManifestFile(icebergTable.io().newInputFile("file:/Users/yzhaoqin/0.12-out.avro"), icebergTable.spec().specId()),
    icebergTable.io());
```

and it worked.

So so far I am not able to reproduce the issue, @stevenzwu @openinx please let me know if you are able to have a repro.

**Comment by stevenzwu** [comment]
I also tried to reproduce the issue and wasn't able to reproduce it.
https://gist.github.com/stevenzwu/62198e32b1174f6ea6d093fdc45b161c

Now I realized that our users also reported some other issue with both 0.12
and 0.13 Iceberg jars in the classpath. Maybe that was also the cause of
the exception I reported in
https://github.com/apache/iceberg/pull/3015/files#r753631713.

In summary, this is a non-issue and we can remove it as a blocker for 0.13.
